### DISCUSSION TITLE
Any new resources for AirScan/eSCL proper communication?

### DISCUSSION THREAD

**Discussion by markosjal** [alexpevzner/sane-airscan#157] [Q&A]
Looking to update an old project at https://github.com/markosjal/AirScan and creating a new one to hopefully share a Windows TWAIN scanner with Linux or OSX by use of eSCL/Airscan protocol  .

Desperately in need of traces of successful scans hopefully like wireshark traces between Mac and an AirScan scanner. 

It seems I am having issues with the messaging especially on OSX.

Any help  appreciated

Thanks
Mark

**Comment by alexpevzner** [comment]
Hi Mark,

Mopria has published eSCL protocol specification:

* https://mopria.org/spec-download

So there is no need to reverse engineer it anomure

**Reply by markosjal** [reply]
Thanks, I did happen to find that recently. 

Also curious do you ever develop any Windows Software? Personally I am not a Windows guy but I have encountered a way to share a Windows TWAIN scanner (not WIA) to  eSCL . If you or anyone who sees this may be interested I could use a second brain and another pair of hands.

All I do is in PHP and of course this works well with eSCL protocol anyway. It is all in PHP but some parts may be better off done differently, and relieve the project of the third party dependencies such as Bonjour from Apple and CmdTwain.

**What you would need:**

PHP install (optionally Apache or other web server but at this point I uuse the built in php web server)

Windows scanner with Working Twain Driver (NOT WIA! to act as host/Server). Windows 7 - Windows 10 should work!

OS X computer or a Linux Computer with Airscan/eSCL protocol (Preferably both, to act as clients)


**How it works:**

Install PHP for Windows (Together with the php code, this will manage all scanner communication between client and server and translate scan commands to/from eSCL to CmdTwain)

Install Bonjour Print Services for Windows (After figuring this out, this will give us the ability to advertise or announce the scanner via Bonjour/Avahi)

Install CmdTwain this allows us to communicate with the Twain scanner interface over a command line that is generated in PHP and passed to the TWAIN interface via this utility.


**Challenges thus far:**

Refining eSCL commands in xml

Translating xml/CmdTwain commands 

Running windows commands as services? ( I am not a windows guy) , such a Java to initiate the Bonjour Advertisement and launching php development server from command line at startup (although this could be Apache or other web server).

**Comment by alexpevzner** [comment]
Yes, I did Windows software development, but for me Windows software is like dishwasher firmware: I can write it, but I won't use Windows as a host, only as target. So I don't have any personal interest in development something for Windows just for fun.

You may consider writing this service in Golang. Go is easy to learn, it has excellent cross-compilation support (so you can use Linux to build Windows executable), it has excellent HTTP support just on its standard library and reasonable level of XML support (enough to handle eSCL). Executables build with Go are statically linked and have literally zero external dependencies. so installation is always trivial.
